src/SubscribeThread.cpp

```diff
@@ -1304,14 +1304,35 @@ void SharedData::subscribe_events()
 
 			try
 			{
-				event_id = sig->attr->subscribe_event(
-												Tango::ARCHIVE_EVENT,
-												sig->archive_cb,
-												/*stateless=*/false);
+                try
+                {
+                    event_id = sig->attr->subscribe_event(
+                                                    Tango::ARCHIVE_EVENT,
+                                                    sig->archive_cb,
+                                                    /*stateless=*/false);
+                }
+                catch (Tango::DevFailed &e)
+                {
+                    if (hdb_dev->SubscribeChangeAsFallback)
```

Comment:
There is a compilation error because of this line.
One fix would be to add
#include "HdbEventSubscriber.h"
At the top of src/SubscribeThread.cpp and to replace this line 1316 with:
`if ((static_cast<HdbEventSubscriber *>(hdb_dev->_device))->subscribeChangeAsFallback)`

Another way would be to add a member in hdb_dev reflecting the value of the property...

Reply:
To be consistent with the rest of the code a member in HdbDevice should be added and initialized passing the device property as a parameter to the HdbDevice constructor.

Reply:
Hi Graziano,

I just did the changes you suggested: https://github.com/tango-controls-hdbpp/hdbpp-es/pull/7/commits/154edce43a90245301a2fdc1fffd6f5c0ba7e702

And tested again that the property works as expected,

Can we merge?

Sergi

Reply:
Hi Sergi, Reynald, @lorenzopivetta 
the code looks OK but I have some other questions on this feature:
the Configuration Manager checks that Archive Events are correctly configured when adding a new attribute (unless you set the "manually pushed events" checkbox); in your scenario, do you expect to have by choice only change events configured or code that manually push only change events?
Shouldn't the Configuration Manager be modified to support Change Events to be consistent with this patch?
Also in libhdbpp-mysql only the parameters of Archive Events are stored when ATTR_CONF_EVENT are received.
So Sergi, I would like to understand if you need that HDB++ supports Change Events by design in all its features or otherwise which is the use case of this patch.
The only case when this could be probably useful in my case is when I have device servers pushing only Change Events manually in the code and I cannot/don't want to recompile them. In this scenario this hack would be sufficient and no other modification needed.

Graziano

Reply:
Thanks Graziano for pointing this out. Indeed, the whole HDB++ was designed to check for consistency. As already discussed, I do not like this patch to the archiver, as it opens to inconsistent behavior because the separation of concerns is no more enforced. If this still needs to get in, all the toolchain needs to be updated in order to provide a complete evidence of the running configuration, including the ConfigurationManager and the Configurator GUI... and, to be honest, still I'd not like the approach. I'm wondering whether a custom developed component would be a better approach, not having to compromise on the main design. Ideally, Sergi, wouldn't be possible to just provide a proper archive event configuration for the producer devices?